[WIP] Scenarios

Argn0 opened the pull request:
collecting data for #1575 

Second part of this will be building the APIs to read data from these tables.


Scenarios:

-Item timings for all heroes and all items that cost at least 2000 gold. Timings are  7.5, 10, 12, 15, 20, 25 and 30 minutes. e.g. Anti-Mage buys BF before minute 12.

-How often a player "pings". How does utilizing the ping feature impact a player's win rate? Does overusing this feature maybe lower team morale?

-The hero's lane role (safe, mid , offlane...). On which lane role does a hero have the best win rate?

-First Blood. What is a team's win rate if they get first blood in a match? How important is getting first blood?

-Courier Kill before minute 3.  What is a team's win rate if they get a courier kill before minute 3. How important is it to make sure your courier doesn't die before minute 3?

-Positivity or Negativity in chat. What is a team's win rate if they all chat positive or negative words before minute 1.  Does using positive words like "gl hf" boost team morale?

howardchung commented on manifest.json:
```diff
@@ -184,5 +184,14 @@
       "exec_mode": "cluster",
       "instances": 1
     }
+    ,
```
put the comma on the above line

howardchung commented on manifest.json:
```diff
@@ -184,5 +184,14 @@
+      "script": "svc/scenarios.js",
+      "watch": true,
+      "ignore_watch": [".git", "node_modules"],
+      "group": "socket",
```
group should be backend

howardchung commented on sql/create_tables.sql:
```diff
@@ -421,6 +421,29 @@ CREATE TABLE IF NOT EXISTS leaderboard_rank (
 );
 
+CREATE TABLE IF NOT EXISTS scenarios (
+  hero smallint,
```
suggest using hero_id for consistency with other tables

howardchung commented on sql/create_tables.sql:
```diff
@@ -421,6 +421,29 @@ CREATE TABLE IF NOT EXISTS leaderboard_rank (
+CREATE TABLE IF NOT EXISTS scenarios (
+  hero smallint,
+  item text,
+  time smallint,
```
suggest using int in case a game ends up going really long

howardchung commented on sql/create_tables.sql:
```diff
@@ -421,6 +421,29 @@ CREATE TABLE IF NOT EXISTS leaderboard_rank (
+  item text,
+  time smallint,
+  pings integer,
+  game_duration smallint,
```
use an int.

Is this field necessary?  Could we just use time instead?

howardchung commented on sql/create_tables.sql:
```diff
@@ -421,6 +421,29 @@ CREATE TABLE IF NOT EXISTS leaderboard_rank (
+
+CREATE TABLE IF NOT EXISTS team_scenarios (
+  scenario   text,
+  is_radiant BOOLEAN,
```
lowercase boolean

howardchung commented on sql/create_tables.sql:
```diff
@@ -421,6 +421,29 @@ CREATE TABLE IF NOT EXISTS leaderboard_rank (
+  lane smallint,
+  games bigint DEFAULT 0,
+  wins bigint DEFAULT 0,
+  CONSTRAINT scenario UNIQUE (hero, item, time),
```
I don't think these named constraints are necessary as long as your only insertion is via upsert on conflicting columns.

howardchung commented on svc/parser.js:
```diff
@@ -60,6 +61,11 @@ function parseProcessor(job, cb) {
     } else {
       console.log('completed parse of match %s', match.match_id);
+      if (true || match.origin === 'scanner') {
+        if (Math.random() >= 0) {
```
make this a config value in config.js

howardchung commented on svc/parser.js:
```diff
@@ -60,6 +61,11 @@ function parseProcessor(job, cb) {
       console.error(err.stack || err);
     } else {
       console.log('completed parse of match %s', match.match_id);
+      if (true || match.origin === 'scanner') {
```
remove the `true`?

howardchung commented on svc/parser.js:
```diff
@@ -60,6 +61,11 @@ function parseProcessor(job, cb) {
       console.log('completed parse of match %s', match.match_id);
+      if (true || match.origin === 'scanner') {
+        if (Math.random() >= 0) {
+          redis.lpush('scenariosQueue', match.match_id);
```
optional: add a monitor for this field to the buildStatus object so we can see if the queue is growing

howardchung commented on svc/scenarios.js:
```diff
@@ -0,0 +1,34 @@
+        const values = Object.keys(row.columns).map(() =>
+          '?');
+        const query = util.format(
+          `INSERT INTO %s (%s) VALUES (%s) ON CONFLICT (%s) DO UPDATE SET wins = ${row.table}.wins + ${row.won ? 1 : 0}, games = ${row.table}.games + 1`,
```
you can use EXCLUDED.wins to avoid the need to interpolate a raw string into the query

howardchung replied:
I suggest parameterizing the row.won ? 1 : 0 part with a `?` to lower risk of sql injection

howardchung replied:
I think you need to set the initial values here for wins and games, otherwise your counts will be off by 1 (the first insert will result in 0 for both fields)

howardchung commented on util/scenariosUtil.js:
```diff
@@ -0,0 +1,143 @@
+        if (player.purchase_log) {
+          player.purchase_log.forEach((item) => {
+            if (dotaItems.indexOf(item.key) !== -1 && item.time <= timings[timings.length - 1]) {
+              const won = (player.player_slot < 5 && match.radiant_win) || (player.player_slot > 4 && !match.radiant_win);
```
There is a utility function called isRadiant that you can use here

howardchung commented on util/scenariosUtil.js:
```diff
@@ -0,0 +1,143 @@
+                columns: {
+                  hero: player.hero_id,
+                  item: item.key,
+                  time: timings.find(x => x >= item.time),
```
you can add the wins column here as a field instead of a separate prop called won.  It will be 1 or 0n and the upsert logic can handle it automatically

howardchung commented on util/scenariosUtil.js:
```diff
@@ -0,0 +1,143 @@
+    if (match.players) {
+      match.players.forEach((player) => {
+        if (player.pings <= pingBucket[pingBucket.length - 1]) {
+          const pings = pingBucket.find(x => x >= player.pings) || -2;
```
what is -2?

howardchung commented on util/scenariosUtil.js:
```diff
@@ -0,0 +1,143 @@
+      match.players.forEach((player) => {
+        if (player.pings <= pingBucket[pingBucket.length - 1]) {
+          const pings = pingBucket.find(x => x >= player.pings) || -2;
+          const won = (player.player_slot < 5 && match.radiant_win) || (player.player_slot > 4 && !match.radiant_win);
```
use isRadiant util function

howardchung commented on util/scenariosUtil.js:
```diff
@@ -0,0 +1,143 @@
+          break;
+        }
+        if (negativeWords.some(word => c.key.toLowerCase().indexOf(word) !== -1)) {
+          if (c.slot < 5) {
```
use the `player_slot` field, then you can also use the utility function here

Argn0 replied:
I think that won't work because the utility function takes a player object and the the chat log is in match.chat

Argn0 commented:
-constraints are no longer named
-first insert now works correctly and no longer results in 0 for both fields
-all query values are now parameterized
TODO: add more scenarios

howardchung commented on sql/migrations.sql:
```diff
@@ -10,3 +10,25 @@ CREATE INDEX IF NOT EXISTS public_matches_avg_rank_tier_idx on public_matches(av
+  item text,
+  time integer,
+  pings integer,
+  lane smallint,
```
we may want to use lane_role instead of lane, since lane is absolute position (top/mid/bot) and lane_role is normalized for side (safe/mid/off)

howardchung commented on svc/scenarios.js:
```diff
@@ -0,0 +1,37 @@
+            table,
+            Object.keys(row).join(','),
+            values,
+            Object.keys(row).slice(0, -1).join(','),
```
what's the purpose of .slice(0, -1) here?

howardchung commented on svc/scenarios.js:
```diff
@@ -0,0 +1,37 @@
+          const values = Object.keys(row).map(() =>
+            '?');
+          const query = util.format(
+            'INSERT INTO %s (%s) VALUES (%s) ON CONFLICT (%s) DO UPDATE SET wins = (%s).wins + EXCLUDED.wins, games = (%s).games + 1',
```
I don't think you need the parentheses for (%s).wins and (%s).games

Argn0 commented:
-replaced slice method with filter method. (wins column is not an ON CONFLICT target)
-removed unnecessary  parentheses in query string
-replaced "lane" with "lane_role"
